File: androidApp/feature-history/src/main/kotlin/dev/roozbahani/trailmetrics/feature/history/DetailsViewModel.kt
package dev.roozbahani.trailmetrics.feature.history

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import dev.roozbahani.trailmetrics.domain.model.ActivityRecord
import dev.roozbahani.trailmetrics.domain.repository.ActivityHistoryRepository
import dev.roozbahani.trailmetrics.feature.history.util.deleteSnapshotFile
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

class DetailsViewModel(
    private val activityId: Long,
    private val activityHistoryRepository: ActivityHistoryRepository
) : ViewModel() {

    private val _uiState = MutableStateFlow(DetailsUiState())
    val uiState: StateFlow<DetailsUiState> = _uiState.asStateFlow()

    init {
        viewModelScope.launch {
            val activity = activityHistoryRepository.getActivity(activityId)
            _uiState.update { it.copy(activity = activity, isLoading = false) }
        }
    }

    fun onDeleteConfirmed(onDeleted: () -> Unit) {
        viewModelScope.launch {
            activityHistoryRepository.deleteActivity(activityId)
            deleteSnapshotFile(_uiState.value.activity?.snapshotFilePath)
            onDeleted()
        }
    }
}

data class DetailsUiState(
    val activity: ActivityRecord? = null,
    val isLoading: Boolean = true
)

File: iosApp/Packages/History/Sources/History/DetailsViewModel.swift
//
//  DetailsViewModel.swift
//  History
//

import Foundation
import SharedKit

@MainActor
public class DetailsViewModel: ObservableObject {
    @Published public var activity: ActivityRecord?
    @Published public var isLoading = true

    private let activityId: Int64
    private let activityHistoryRepository: ActivityHistoryRepository

    public init(
        activityId: Int64,
        activityHistoryRepository: ActivityHistoryRepository = KoinHelper().getActivityHistoryRepository()
    ) {
        self.activityId = activityId
        self.activityHistoryRepository = activityHistoryRepository

        Task {
            activity = try? await activityHistoryRepository.getActivity(id: activityId)
            isLoading = false
        }
    }
}
